man1/ael.1
'\" t
.\"	aegis - project change supervisor
.\"	Copyright (C) 1991, 1992, 1993, 1994 Peter Miller.
.\"	All rights reserved.
.\"
.\"	This program is free software; you can redistribute it and/or modify
.\"	it under the terms of the GNU General Public License as published by
.\"	the Free Software Foundation; either version 2 of the License, or
.\"	(at your option) any later version.
.\"
.\"	This program is distributed in the hope that it will be useful,
.\"	but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"	GNU General Public License for more details.
.\"
.\"	You should have received a copy of the GNU General Public License
.\"	along with this program; if not, write to the Free Software
.\"	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" MANIFEST: manual entry for 'aegis -List' command
.\"
.so z_name.so
.TH "\*(n) -List" 1 \*(N)
.SH NAME
\*(n) -List \- list (possibly) interesting things
.SH SYNOPSIS
.B \*(n)
.B -List
[
.IR option ...
]
.I list-name
.br
.B \*(n)
.B -List
.B -List
[
.IR option ...
]
.br
.B \*(n)
.B -List
.B -Help
.SH DESCRIPTION
The
.I \*(n)
.I -List
command is used to
list information.
There are a number of possible
.IR list-name s,
as follows (abbreviations as for command line options):
.TP 8n
Administrators
.br
List the administrators of a project.
.TP 8n
Change_Details
.br
List full information about a change
in large format.
.TP 8n
Change_Files
.br
List all files in a change.
.TP 8n
Change_History
.br
List the history of a change.
.TP 8n
Changes
.br
List the changes of a project.
.TP 8n
Default_Change
.br
List the default change for the current user.
.TP 8n
Default_Project
.br
List the default project for the current user.
.TP 8n
Developers
.br
List the developers of a project.
.TP 8n
Integrators
.br
List the integrators of a project.
.TP 8n
List_List
.br
List all lists available.
.TP 8n
Locks
.br
List all currently active locks.
.TP 8n
Outstanding_Changes
.br
List all changes not yet completed.
.TP 8n
All_Outstanding_Changes
.br
List all changes not yet completed, for all projects.
.TP 8n
Project_Files
.br
List all files in the baseline of a project.
.TP 8n
Project_History
.br
List the integration history of a project.
.TP 8n
Projects
.br
List all projects.
.TP 8n
Reviewers
.br
List the reviewers of a project.
.TP 8n
Users_Changes
.br
List of changes owned by the current user.
.TP 8n
Version
.br
List version of a project or change.
This includes the major and minor version number,
and the previous version number if available.
The list of copyright years is also printed.
.PP
Most of these lists are available from other \*(n) functions.
Many \*(n) functions provide more specific lists.
.SH OPTIONS
The following options are understood:
.so o_change.so
.so o_help.so
.so o_list.so
.so o_page.so
.so o_project.so
.so o_terse.so
.so o_unforma.so
.so o_verbose.so
.so o__rules.so
.SH RECOMMENDED ALIAS
The recommended alias for this command is
.nf
.ta 8n 16n
csh%	alias ael '\*(n) -l \e!* -v'
sh$	ael(){\*(n) -l $* -v}
.fi
.SH ERRORS
It is an error if
the list name given is unknown.
.so z_exit.so
.br
.ne 1i
.SH SEE ALSO
.TP 8n
.IR aedb (1)
begin development of a change (listing option)
.TP 8n
.IR aeib (1)
begin integration of a change (listing option)
.TP 8n
.IR aer (1)
report generator
.TP 8n
.IR aerp (1)
pass review of a change (listing option)
.TP 8n
.IR aeuconf (5)
user configuration file format
.so z_cr.so


etc/CHANGES



Project "aegis"								  Page 1
Details							Sat Apr	 5 23:36:53 2003

NAME
	Project "aegis".

SUMMARY
	project change supervisor.

DESCRIPTION
	project change supervisor.

BRANCH CONTENTS
	Change	State		Description
	------- -------		-----------
	3	completed	project change supervisor, branch 3.
	4	being_developed project change supervisor, branch 4.

ARCHITECTURES
	This change must be developed on the unspecified architectures.

	arch.	 host	  aeb	      aet	  aet -bl     aet -reg
	-------- -------- ---------   ---------	  ---------   ---------
	unspecif	  required    required	  required    exempt
	ied

CAUSE
	This change was caused by internal_enhancement.

FILES
	Type	Action	 Edit	      File Name
	------	------	 ------	      ------
	build	remove	 5	      BUILDING
				      Deleted by change 3.
	source	remove	 2 (5)	      LICENSE
				      Deleted by change 3.
	build	remove	 8 (5)	      MANIFEST
				      Deleted by change 3.
	build	remove	 10 (5)	      Makefile.in
				      Deleted by change 3.
	build	remove	 6 (5)	      README
				      Deleted by change 3.
	source	create	 7 (5)	      aedist/arglex3.c
	source	create	 7 (5)	      aedist/arglex3.h
	source	remove	 4 (5)	      aedist/header.c
				      Deleted by change 3.
				      Moved to libaegis/rfc822header.c
	source	remove	 1 (5)	      aedist/header.h
				      Deleted by change 3.
				      Moved to libaegis/rfc822header.h
	source	remove	 5	      aedist/input/822wrap.c
				      Deleted by change 3.
				      Moved to libaegis/input/822wrap.c
	source	remove	 2 (5)	      aedist/input/822wrap.h
				      Deleted by change 3.
				      Moved to libaegis/input/822wrap.h
	source	create	 3 (5)	      aedist/input/cpio.c
	source	create	 1 (5)	      aedist/input/cpio.h
	source	create	 5	      aedist/input/cpio_child.c
	source	create	 1 (5)	      aedist/input/cpio_child.h



Project "aegis"								  Page 2
Details							Sat Apr	 5 23:36:53 2003

	Type	Action	 Edit	      File Name
	------	------	 ------	      ------
	source	create	 9 (5)	      aedist/list.c
	source	create	 1 (5)	      aedist/list.h
	source	create	 5	      aedist/main.c
	source	create	 7 (5)	      aedist/open.c
	source	create	 2 (5)	      aedist/open.h
	source	create	 4 (5)	      aedist/output/cpio.c
	source	create	 1 (5)	      aedist/output/cpio.h
	source	create	 5	      aedist/output/cpio_child.c
	source	create	 1 (5)	      aedist/output/cpio_child.h
	source	create	 4 (5)	      aedist/output/cpio_child2.c
	source	create	 1 (5)	      aedist/output/cpio_child2.h
	source	create	 29 (5)	      aedist/receive.c
	source	create	 1 (5)	      aedist/receive.h
	source	create	 21 (5)	      aedist/send.c
	source	create	 1 (5)	      aedist/send.h
	source	create	 2 (5)	      aefind/cmdline.h
	source	create	 8 (5)	      aefind/cmdline.y
	source	create	 4 (5)	      aefind/descend.c
	source	create	 2 (5)	      aefind/descend.h
	source	create	 3 (5)	      aefind/function.c
	source	create	 2 (5)	      aefind/function.h
	source	create	 2 (5)	      aefind/function/basename.c
	source	create	 2 (5)	      aefind/function/basename.h
	source	create	 3 (5)	      aefind/function/needs.c
	source	create	 2 (5)	      aefind/function/needs.h
	source	create	 2 (5)	      aefind/function/print.c
	source	create	 2 (5)	      aefind/function/print.h
	source	create	 4 (5)	      aefind/function/stat.c
	source	create	 2 (5)	      aefind/function/stat.h
	source	create	 5	      aefind/lex.c
	source	create	 2 (5)	      aefind/lex.h
	source	create	 2 (5)	      aefind/main.c
	source	create	 2 (5)	      aefind/shorthand/name.c
	source	create	 2 (5)	      aefind/shorthand/name.h
	source	create	 2 (5)	      aefind/shorthand/path.c
	source	create	 2 (5)	      aefind/shorthand/path.h
	source	create	 2 (5)	      aefind/shorthand/print.c
	source	create	 2 (5)	      aefind/shorthand/print.h
	source	create	 3 (5)	      aefind/shorthand/stat.c
	source	create	 2 (5)	      aefind/shorthand/stat.h
	source	create	 2 (5)	      aefind/tree.c
	source	create	 2 (5)	      aefind/tree.h
	source	create	 4 (5)	      aefind/tree/arithmetic.c
	source	create	 2 (5)	      aefind/tree/arithmetic.h
	source	create	 3 (5)	      aefind/tree/bitwise.c
	source	create	 2 (5)	      aefind/tree/bitwise.h
	source	create	 2 (5)	      aefind/tree/constant.c
	source	create	 2 (5)	      aefind/tree/constant.h
	source	create	 2 (5)	      aefind/tree/diadic.c
	source	create	 2 (5)	      aefind/tree/diadic.h
	source	create	 2 (5)	      aefind/tree/list.c
	source	create	 2 (5)	      aefind/tree/list.h
	source	create	 3 (5)	      aefind/tree/logical.c
	source	create	 2 (5)	      aefind/tree/logical.h
	source	create	 3 (5)	      aefind/tree/match.c



Project "aegis"								  Page 3
Details							Sat Apr	 5 23:36:53 2003

	Type	Action	 Edit	      File Name
	------	------	 ------	      ------
	source	create	 2 (5)	      aefind/tree/match.h
	source	create	 2 (5)	      aefind/tree/monadic.c
	source	create	 2 (5)	      aefind/tree/monadic.h
	source	create	 2 (5)	      aefind/tree/now.c
	source	create	 2 (5)	      aefind/tree/now.h
	source	create	 2 (5)	      aefind/tree/private.c
	source	create	 2 (5)	      aefind/tree/private.h
	source	create	 4 (5)	      aefind/tree/relative.c
	source	create	 2 (5)	      aefind/tree/relative.h
	source	create	 2 (5)	      aefind/tree/this.c
	source	create	 2 (5)	      aefind/tree/this.h
	source	create	 3 (5)	      aefp/main.c
	source	create	 17 (5)	      aegis/aeb.c
	source	create	 2 (5)	      aegis/aeb.h
	source	create	 13 (5)	      aegis/aeca.c
	source	create	 3 (5)	      aegis/aeca.h
	source	create	 10 (5)	      aegis/aecd.c
	source	create	 2 (5)	      aegis/aecd.h
	source	create	 11 (5)	      aegis/aechown.c
	source	create	 2 (5)	      aegis/aechown.h
	source	create	 12 (5)	      aegis/aeclean.c
	source	create	 1 (5)	      aegis/aeclean.h
	source	create	 19 (5)	      aegis/aeclone.c
	source	create	 1 (5)	      aegis/aeclone.h
	source	create	 19 (5)	      aegis/aecp.c
	source	create	 2 (5)	      aegis/aecp.h

aegis/ael.cc
//
// aegis - project change supervisor
// Copyright (C) 1991-1999, 2001-2008, 2012 Peter Miller
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or (at
// your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
//
// If you add another list to this file,
// don't forget to update man1/ael.1
//

#include <common/ac/stdio.h>

#include <common/progname.h>
#include <common/quit.h>
#include <common/sizeof.h>
#include <common/str_list.h>
#include <common/trace.h>
#include <libaegis/ael/branch/files.h>
#include <libaegis/ael/change/changes.h>
#include <libaegis/ael/change/default.h>
#include <libaegis/ael/change/details.h>
#include <libaegis/ael/change/file_history.h>
#include <libaegis/ael/change/file_invento.h>
#include <libaegis/ael/change/files.h>
#include <libaegis/ael/change/history.h>
#include <libaegis/ael/change/inappropriat.h>
#include <libaegis/ael/change/incomplete.h>
#include <libaegis/ael/change/inventory.h>
#include <libaegis/ael/change/outstand_all.h>
#include <libaegis/ael/change/outstanding.h>
#include <libaegis/ael/change/user.h>
#include <libaegis/ael/change/version.h>
#include <libaegis/ael/column_width.h>
#include <libaegis/ael/formeditnum.h>
#include <libaegis/ael/locks.h>
#include <libaegis/ael/project/administrato.h>
#include <libaegis/ael/project/aliases.h>
#include <libaegis/ael/project/default.h>
#include <libaegis/ael/project/details.h>
#include <libaegis/ael/project/developers.h>
#include <libaegis/ael/project/file_invento.h>
#include <libaegis/ael/project/filesbydelta.h>
#include <libaegis/ael/project/files.h>
#include <libaegis/ael/project/history.h>
#include <libaegis/ael/project/inappropriat.h>
#include <libaegis/ael/project/integrators.h>
#include <libaegis/ael/project/projects.h>
#include <libaegis/ael/project/reviewers.h>
#include <libaegis/ael/state_file.h>
#include <libaegis/arglex/change.h>
#include <libaegis/arglex/project.h>
#include <libaegis/arglex2.h>
#include <libaegis/change/identifier.h>
#include <libaegis/col.h>
#include <libaegis/help.h>
#include <libaegis/option.h>
#include <libaegis/output.h>
#include <libaegis/sub.h>
#include <libaegis/zero.h>

#include <aegis/ael.h>


static void list_list_list(change_identifier &, string_list_ty *); // forward


struct table_ty
{
    const char *name;
    const char *description;
    void (*func)(change_identifier &, string_list_ty *);
    int number_allowed_args;
};


static table_ty table[] =
{
    {
        "Administrators",
        "List the administrators of a project",
        list_administrators,
        0,
    },
    {
        "Branch_Files",
        "List the files in a branch",
        list_branch_files,
        0,
    },
    {
        "Change_Details",
        "List all information about a change in large layout form.",
        list_change_details,
        0,
    },
    {
        "Change_Files",
        "List all files in a change",
        list_change_files,
        0,
    },
    {
        "Change_File_History",
        "List the history of all files in a change.",
        list_change_file_history,
        0,
    },
    {
        "Change_File_INventory",
        "List all files in a change with their UUID.",
        list_change_file_inventory,
        0,
    },
    {
        "Change_History",
        "List the history of a change",
        list_change_history,
        0,
    },
    {
        "Change_INventory",
        "List the changes of a project with their UUID.",
        list_change_inventory,
        0,
    },
    {
        "Changes",
        "List the changes of a project",
        list_changes,
        0,
    },
    {
        "Default_Change",
        "List the default change for the current user",
        list_default_change,
        1,
    },
    {
        "Default_Project",
        "List the default project for a (the current) user",
        list_default_project,
        1,
    },
    {
        "Developers",
        "List the developers of a project",
        list_developers,
        0,
    },
    {
        "INComplete",
        "List the incomplete changes (for a user)",
        list_incomplete_changes,
        0,
    },
    {
        "Integrators",
        "List the integrators of a project",
        list_integrators,
        0,
    },
    {
        "List_List",
        "List all lists available",
        list_list_list,
        0,
    },
    {
        "Locks",
        "List the active locks",
        list_locks,
        0,
    },
    {
        "Outstanding_Changes",
        "List the outstanding changes (for a user)",
        list_outstanding_changes,
        1,
    },
    {
        "All_Outstanding_Changes",
        "List the outstanding changes for all projects",
        list_outstanding_changes_all,
        0,
    },
    {
        "Branch_Details",
        "List the details of all changes in the branch (recursive)",
        list_project_details,
        0,
    },
    {
        "Project_Details",
        "List the details of all changes in the project (recursive)",
        list_project_details,
        0,
    },
    {
        "Project_Files",
        "List all files in the baseline of a project",
        list_project_files,
        0,
    },
    {
        "Project_Files_By_Delta",
        "List all files in the baseline of a project, as they appeared "
            "immediately after the integrate pass of the specified "
            "change or delta.",
        list_project_files_by_delta,
        0,
    },
    {
        "Project_File_INventory",
        "List all files in the baseline of a project with their UUID.",
        list_project_file_inventory,
        0,
    },
    {
        "Project_History",
        "List the integration history of a project",
        list_project_history,
        0,
    },
    {
        "Projects",
        "List all projects",
        list_projects,
        0,
    },
    {
        "Project_Aliases",
        "List all project aliases (the -Project option limits results)",
        list_project_aliases,
        0,
    },
    {
        "Reviewers",
        "List the reviewers of a project",
        list_reviewers,
        0,
    },
    {
        "State_File_Name",
        "List the full path name of the change's state file.",
        list_state_file_name,
        0,
    },
    {
        "Users_Changes",
        "List of changes owned by a (the current) user",
        list_user_changes,
        1,
    },
    {
        "Version",
        "List version of a project or change",
        list_version,
        0,
    },
};


static void
list_usage(void)
{
    const char      *progname;

    progname = progname_get();
    fprintf(stderr, "usage: %s -List [ <option>... ] <listname>\n", progname);
    fprintf(stderr, "       %s -List -List [ <option>... ]\n", progname);
    fprintf(stderr, "       %s -List -Help\n", progname);
    quit(1);
}


static void
list_help(void)
{
    help("ael", list_usage);
}


static void
list_list(void)
{
    trace(("list_list()\n{\n"));
    arglex();
    change_identifier cid;
    while (arglex_token != arglex_token_eoln)
        generic_argument(list_usage);
    list_list_list(cid, 0);
    trace(("}\n"));
}


static int
max_num_table_args(void)
{
    size_t          i;
    int             max_num_args = 0;
    trace(("max_num_table_args()\n{\n"));
    for (i = 0; i < SIZEOF(table); i++)
    {
        if (table[i].number_allowed_args > max_num_args)
        {
            max_num_args = table[i].number_allowed_args;
        }
    }
    trace(("\tmax_num_args = %d\n", max_num_args));
    trace(("}\n"));
    return max_num_args;
}


static void
list_main(void)
{
    sub_context_ty  *scp;
    table_ty        *tp;
    int             j;
    table_ty        *hit[SIZEOF(table)];
    int             nhit;
    string_ty       *s1;
    string_ty       *s2;

    trace(("list_main()\n{\n"));
    arglex();
    const char *listname = 0;
    change_identifier cid;
    string_list_ty args_list;
    while (arglex_token != arglex_token_eoln)
    {
        switch (arglex_token)
        {
        default:
            generic_argument(list_usage);
            continue;

        case arglex_token_string:
            if (listname)
            {
                string_ty       *s;

                s = str_from_c(arglex_value.alv_string);
                args_list.push_back(s);
                str_free(s);

                if (args_list.nstrings > (size_t)max_num_table_args())
                {
                    fatal_intl(0, i18n("too many lists"));
                }
            }
            else
            {
                listname = arglex_value.alv_string;
            }
            break;

        case arglex_token_baseline:
        case arglex_token_branch:
        case arglex_token_change:
        case arglex_token_delta:
        case arglex_token_delta_date:
        case arglex_token_development_directory:
        case arglex_token_grandparent:
        case arglex_token_number:
        case arglex_token_project:
        case arglex_token_trunk:
            cid.command_line_parse(list_usage);
            continue;
        }
        arglex();
    }
    cid.command_line_check(list_usage);
    if (!listname)
        fatal_intl(0, i18n("no list"));

    nhit = 0;
    for (tp = table; tp < ENDOF(table); ++tp)
    {
        if (arglex_compare(tp->name, listname, 0))
            hit[nhit++] = tp;
    }
    switch (nhit)
    {
    case 0:
        scp = sub_context_new();
        sub_var_set_charstar(scp, "Name", listname);
        fatal_intl(scp, i18n("no $name list"));
        // NOTREACHED
        sub_context_delete(scp);

    case 1:
        hit[0]->func(cid, &args_list);
        break;

    default:
        s1 = str_from_c(hit[0]->name);
        for (j = 1; j < nhit; ++j)
        {
            s2 = str_format("%s, %s", s1->str_text, hit[j]->name);
            str_free(s1);
            s1 = s2;
        }
        scp = sub_context_new();
        sub_var_set_charstar(scp, "Name", listname);
        sub_var_set_string(scp, "Name_List", s1);
        str_free(s1);
        sub_var_optional(scp, "Name_List");
        fatal_intl(scp, i18n("list $name ambiguous"));
        // NOTREACHED
        sub_context_delete(scp);
        break;
    }
    trace(("}\n"));
}


void
list(void)
{
    static arglex_dispatch_ty dispatch[] =
    {
        { arglex_token_help, list_help, 0 },
        { arglex_token_list, list_list, 0 },
    };

    trace(("list()\n{\n"));
    arglex_dispatch(dispatch, SIZEOF(dispatch), list_main);
    trace(("}\n"));
}


static void
list_list_list(change_identifier &cid, string_list_ty *)
{
    trace(("list_list_list()\n{\n"));
    if (cid.project_set())
        list_project_inappropriate();
    if (cid.set())
        list_change_inappropriate();

    //
    // create the columns
    //
    col::pointer colp = col::open((string_ty *)0);
    colp->title("List of Lists", (const char *)0);
    output::pointer name_col = colp->create(0, 15, "Name\n------");
    output::pointer desc_col;
    if (!option_terse_get())
    {
        desc_col = colp->create(16, 0, "Description\n-------------");
    }

    //
    // list the lists
    //
    for (table_ty *tp = table; tp < ENDOF(table); ++tp)
    {
        name_col->fputs(tp->name);
        if (desc_col)
            desc_col->fputs(tp->description);
        colp->eoln();
    }
    trace(("}\n"));
}


// vim: set ts=8 sw=4 et :
